### DISCUSSION TITLE
Please add option 'Turn the depthmap into absolute black/white'

### DISCUSSION THREAD

**Discussion by G-force78** [thygate/stable-diffusion-webui-depthmap-script#69] [Ideas]
Firstly thank you so much for fixing the depth map issue I was going out of my mind trying so many fixes.

This is the feature I refer to  https://github.com/Extraltodeus/depthmap2mask
Its a great feature to use to create clean cut backgrounds

**Comment by graemeniedermayer** [comment]
One thing to consider is rather than using a hard cut on depth. We could use the rembg repo, https://github.com/danielgatis/rembg . it can sometimes help the depth information.

Current depth calculation

![03380-3779724257-_depth](https://user-images.githubusercontent.com/38518120/210177135-5946141f-fceb-409d-bc55-c0baeb78ee94.png)

background removal before depth calculation

![03378-2667539560-_depth](https://user-images.githubusercontent.com/38518120/210177130-0f4e9663-9fd4-4e86-8ac6-9d1e5a54971d.png)

Reapplying the mask after depth calculation.

![depthMaskWithBackgroundRemoved](https://user-images.githubusercontent.com/38518120/210177308-a226dec9-5b5b-44ba-afa6-33ccf2a7bb66.png)

It can be really useful for making layered 3d scenes. I did start are extension for rembg (I haven't fully finished testing it although it might work for you). Initially I was interested in the clothing segmentation so that's why it isn't called rembg. 
https://github.com/graemeniedermayer/clothseg 

I could probably put together a pull request in the next few days if it's of interest?

**Reply by graemeniedermayer** [reply]
After rereading the original post, I should add that you can also turn these can also be turned into black/white masks too.

**Comment by thygate** [comment]
The next version has options to clip and expand depth range with near and far sliders, set them equal to get a b&w cut mask ..
(The 3D-Photo-Inpainting is also in the next version, no options yet, and i still need to do some more testing, it's coming soon)

The near and far clipping works on the 16 bit depth map and gives you 1000 steps on each slider to select the threshold.

Of course it's a bit cumbersome finding the threshold to cut at (0.8 is a good start) by regenerating the map constantly ..
Preferably this should be done client side in js to edit already generated depth maps, and have an option like a 'colorpicker' or histogram to more easily find the depth to cut at. I have no plans for this in the near future though ..

**Reply by thygate** [reply]
I made the commit, feature is now available.

**Reply by steampunkchris** [reply]
Quick question... are all the depth maps currently saved in 16 bit?

edit: LOL - answered my own question... Yes! They are... I am very excited about that =D

**Comment by thygate** [comment]
Is the clip option adequate to generate the b&w masks, create clean cut backgrounds ?

**Reply by G-force78** [reply]
> Is the clip option adequate to generate the b&w masks, create clean cut backgrounds ?

Im not too sure I am getting another error now

```
Error loading script: depthmap.py
Traceback (most recent call last):
  File "/content/stable-diffusion-webui/modules/scripts.py", line 205, in load_scripts
    module = script_loading.load_module(scriptfile.path)
  File "/content/stable-diffusion-webui/modules/script_loading.py", line 13, in load_module
    exec(compiled, module.__dict__)
  File "/content/stable-diffusion-webui/extensions/stable-diffusion-webui-depthmap-script/scripts/depthmap.py", line 54, in <module>
    from inpaint.mesh import write_ply, read_ply, output_3d_photo
  File "/content/stable-diffusion-webui/extensions/stable-diffusion-webui-depthmap-script/scripts/inpaint/mesh.py", line 16, in <module>
    from moviepy.editor import ImageSequenceClip
  File "/usr/local/lib/python3.8/dist-packages/moviepy/editor.py", line 26, in <module>
    imageio.plugins.ffmpeg.download()
  File "/usr/local/lib/python3.8/dist-packages/imageio/plugins/ffmpeg.py", line 37, in download
    raise RuntimeError(
RuntimeError: imageio.ffmpeg.download() has been deprecated. Use 'pip install imageio-ffmpeg' instead
```

**Reply by thygate** [reply]
issue with imageio, please run `pip install imageio==2.4.1`

**Reply by thygate** [reply]
well, any progress ?

**Comment by G-force78** [comment]
It does make a nice cut however for some reason the inpainting is only picking up the outside of the mask, which could well be because of the model Im using at the  moment which is a custom dreambooth one so will have to test it on the 1.5inpaint or sd 2.1. From top to bottom after the original init image is far 0 near 0.3, far 0 near 0.6, far 0 near 0.9.
![00011-603354621-kneeling cat knight, portrait, finely detailed armor, intricate design, silver, silk, cinematic lighting, 4k](https://user-images.githubusercontent.com/114336644/211292297-1a2be94c-ec22-4103-a923-1a3e4ef25e80.png)

![f0n03](https://user-images.githubusercontent.com/114336644/211292302-b7a7156b-133d-4722-aa85-ecaec0c3532f.png)
![f0n06](https://user-images.githubusercontent.com/114336644/211292303-cefb9362-ee9f-4953-bb81-b0cf8b861d38.png)
![f0n09](https://user-images.githubusercontent.com/114336644/211292306-38ea2f33-6f5c-4b94-841a-97bc584318eb.png)
![catknightinpainttest](https://user-images.githubusercontent.com/114336644/211292848-4dcea730-51d0-4247-aa02-75a46bd08eb5.png)

**Comment by thygate** [comment]
Those look like pretty good masks ... Might benefit from a slight post blur maybe ? 

I will also merge the PR by @graemeniedermayer to include the rembg asap ..

**Reply by thygate** [reply]
Merge complete, rembg now available. Would this all benefit from a small blur ?

**Reply by G-force78** [reply]
yeah the masks are good, I dont know what the blur would be for? For the immediate area near the border of the mask? I guess rembg can make a cleaner boundary
